### PR TITLE
Always return absolute path in path_get_cdpath

### LINKED ISSUE #6220: successive CDPATH CD's followed by ".." cause Backtrace SIGABRT termination
fish, version 3.0.2

sh -c 'env HOME=$(mktemp -d) fish'
mkdir .config
cd .config
cd .config
..&lt;E&gt; fish: /builddir/fish-3.0.2/src/wutil.cpp:464: failed assertion: !wd.empty() && wd.front() == sep && wd.ba
ck() == sep && "Invalid working directory, it must start and end with /"
<E> fish: Backtrace:
<E> fish: 0   path_normalize_for_cd(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&) + 1263
<E> fish: 1   fish(+0xee8c4) [0x55b0b92f48c4]
<E> fish: 2   path_can_be_implicit_cd(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >*, env_vars_snapshot_t const&) + 130
<E> fish: 3   fish(+0xbc5eb) [0x55b0b92c25eb]
<E> fish: 4   highlight_shell(std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > const&, std::vector<unsigned int, std::allocator<unsigned int> >&, unsigned long, std::vector<std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> >, std::allocator<std::__cxx11::basic_string<wchar_t, std::char_traits<wchar_t>, std::allocator<wchar_t> > > >*, env_vars_snapshot_t const&) + 425
<E> fish: 5   fish(+0xf96f0) [0x55b0b92ff6f0]
<E> fish: 6   std::_Function_handler<void (), _iothread_trampoline<highlight_result_t>::perform<std::function<highlight_result_t ()>, void (*)(highlight_result_t)>(std::function<highlight_result_t ()> const&, void (* const&)(highlight_result_t))::{lambda()#1}>::_M_invoke(std::_Any_data const&) + 49
<E> fish: 7   fish(+0xdb5e4) [0x55b0b92e15e4]
<E> fish: 8   /usr/lib/libpthread.so.0(+0x8f27) [0x7f15a07aff27]
<E> fish: 9   clone + 63
fish: Job 2, “bash -norc $argv” terminated by signal SIGABRT (Abort)

Typing ".." after the second "cd .config" causes the immediate backtrace. Appears only to occur with path changes using the CDPATH spec e.g. explicit pathnames "cd ~/.config; cd ~/.config; .." does not cause a backtrace. Directory names do not need to be the same to cause the backtrace. 

Occurs on..

Voidlinux: Linux lumen 5.2.21_1 #1 SMP PREEMPT Fri Oct 11 18:24:40 UTC 2019 x86_64 GNU/Linux

Archlinux: Linux luna 5.3.5-arch1-1-ARCH #1 SMP PREEMPT Mon Oct 7 19:03:08 UTC 2019 x86_64 GNU/Linux

Terminals alacritty and xterm: xterm-256color

CDPATH set as..

set -x CDPATH . .. ../..

Not a hugely serious workflow issue, as successive relative CD's followed by ".." would be rare! but the backtrace would be annoying.

### PR DESCRIPTION
**Pull request by krobelus**
Fixes #6220

----

## TODOs:

- [x] Tests have been added for regressions fixed
- [x] User-visible changes noted in CHANGELOG.md

### CODE REVIEW

File: src/path.cpp [hunk 1]
```diff
@@ -181,14 +180,16 @@ maybe_t<wcstring> path_get_cdpath(const wcstring &dir, const wcstring &wd,
...
                 // next_path is just '.', and we have a working directory, so use the wd instead.
                 next_path = wd;
             }
 
-            // If next_path starts with ./ we need to replace the . with the wd.
-            if (string_prefixes_string(L"./", next_path) && !wd.empty()) {
+            // We want to return an absolute path (see issue 6220)
+            if (string_prefixes_string(L"./", next_path)) {
                 next_path = next_path.replace(0, 2, wd);
+            } else if (string_prefixes_string(L"../", next_path) || next_path == L"..") {
```
Comment (faho) [inline src/path.cpp]: Does this also work with `../../`?

Comment (krobelus) [reply src/path.cpp]: Looks like it fails with `../..`. I thought `normalize_path` handles that.

Comment (krobelus) [reply src/path.cpp]: actually it works fine, I was on a different branch.
This just needs to make the path absolute, the rest is resolved by `normalize_path`.

File: src/path.h [hunk 2]
```diff
@@ -59,7 +59,7 @@ wcstring_list_t path_get_paths(const wcstring &cmd, const environment_t &vars);
 ///
 /// \param dir The name of the directory.
 /// \param out_or_NULL If non-NULL, return the path to the resolved directory
```
Comment (ridiculousfish) [inline src/path.h]: while you're here perhaps clean up this stale comment which I forgot to do

**Pull request merged**